libeos-parental-controls/app-filter.c

```diff
@@ -452,3 +483,510 @@ epc_get_app_filter_finish (GAsyncResult  *result,
+  set_app_filter_complete (connection, task);
+}
+
+static void
+set_oars_filter_cb (GObject      *obj,
+                    GAsyncResult *result,
+                    gpointer      user_data)
+{
+  GDBusConnection *connection = G_DBUS_CONNECTION (obj);
+  g_autoptr(GTask) task = G_TASK (user_data);
+
+  SetAppFilterData *data = g_task_get_task_data (task);
+  g_assert (data->set_oars_filter_result == NULL);
+  data->set_oars_filter_result = g_object_ref (result);
+  set_app_filter_complete (connection, task);
+}
+
+static void
+set_app_filter_complete (GDBusConnection *connection,
+                         GTask           *task)
+{
+  g_autoptr(GVariant) app_filter_result_variant = NULL;
+  g_autoptr(GVariant) oars_filter_result_variant = NULL;
+  g_autoptr(GError) app_filter_error = NULL;
+  g_autoptr(GError) oars_filter_error = NULL;
+
+  SetAppFilterData *data = g_task_get_task_data (task);
+
+  if (data->set_app_filter_result == NULL ||
+      data->set_oars_filter_result == NULL)
+    return;
+
+  app_filter_result_variant = g_dbus_connection_call_finish (connection,
+                                                             data->set_app_filter_result,
+                                                             &app_filter_error);
+  oars_filter_result_variant = g_dbus_connection_call_finish (connection,
+                                                              data->set_oars_filter_result,
+                                                              &oars_filter_error);
+
+  if (app_filter_error != NULL)
+    {
+      g_task_return_error (task, bus_error_to_app_filter_error (app_filter_error,
+                                                                data->user_id));
+      return;
+    }
+  if (oars_filter_error != NULL)
+    {
+      g_task_return_error (task, bus_error_to_app_filter_error (oars_filter_error,
+                                                                data->user_id));
+      return;
+    }
+
+  g_task_return_boolean (task, TRUE);
+}
+
+/**
+ * epc_set_app_filter_finish:
+ * @result: a #GAsyncResult
+ * @error: return location for a #GError, or %NULL
+ *
+ * Finish an asynchronous operation to set the app filter for a user, started
+ * with epc_set_app_filter_async().
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ * Since: 0.1.0
+ */
+gboolean
+epc_set_app_filter_finish (GAsyncResult  *result,
+                           GError       **error)
+{
+  g_return_val_if_fail (g_task_is_valid (result, NULL), FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  return g_task_propagate_boolean (G_TASK (result), error);
+}
+
+/*
+ * Actual implementation of #EpcAppFilterBuilder.
+ *
+ * All members are %NULL if un-initialised, cleared, or ended.
+ */
+typedef struct
+{
+  GPtrArray *paths_blacklist;  /* (nullable) (owned) (element-type filename) */
+  GHashTable *oars;  /* (nullable) (owned) (element-type utf8 EpcAppFilterOarsValue) */
```

Comment:
I don't think this should be (nullable). `_end` assumes it is non-NULL (without asserting); all the initialization functions ensure it is non-NULL.

Reply:
I marked it as `(nullable)` because it is `NULL` when cleared (and being cleared is a valid state for a heap-allocated `EpcAppFilterBuilder`). There could be some finessing though.

Note that I didn’t bother adding asserts for `oars` anywhere — just for `paths_blacklist`, since they’re set to `NULL`/non-`NULL` identically.

Reply:
Added some more precondition assertions.